### PR TITLE
Fix: Support model parallelism in HF transformer

### PR DESCRIPTION
**Pull request by gavrissh**
**What this PR does / why we need it**:

Include device_map changes and set it to "auto" to allow model parallelism prevent from OOM trying to fit model into a single device

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

Tested with large models like in llama2 70B

**Checklist**:

- [ ] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

### CODE REVIEW

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 1]
```diff
@@ -108,31 +109,42 @@ def load(self) -> bool:
...
 
+        model_config = AutoConfig.from_pretrained(model_id_or_path)
+
         if not self.task:
-            self.task = self.infer_task_from_model_architecture(model_id_or_path)
+            self.task = self.infer_task_from_model_architecture(model_config)
         # load huggingface tokenizer
         self.tokenizer = AutoTokenizer.from_pretrained(
-            model_id_or_path, do_lower_case=self.do_lower_case)
+            model_id_or_path, do_lower_case=self.do_lower_case, device_map="auto")
```
Comment (terrytangyuan) [inline python/huggingfaceserver/huggingfaceserver/model.py]: Consider moving these to constants?

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 2]
```diff
@@ -108,31 +109,42 @@ def load(self) -> bool:
...
                 return True
 
+        model_config = AutoConfig.from_pretrained(model_id_or_path)
+
         if not self.task:
-            self.task = self.infer_task_from_model_architecture(model_id_or_path)
+            self.task = self.infer_task_from_model_architecture(model_config)
         # load huggingface tokenizer
         self.tokenizer = AutoTokenizer.from_pretrained(
-            model_id_or_path, do_lower_case=self.do_lower_case)
```
Comment (terrytangyuan) [inline python/huggingfaceserver/huggingfaceserver/model.py:115]: Reuse `self.device_map`?

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 3]
```diff
@@ -108,31 +109,43 @@ def load(self) -> bool:
                 self.vllm_engine = AsyncLLMEngine.from_engine_args(self.vllm_engine_args)
                 return True
 
+        model_config = AutoConfig.from_pretrained(model_id_or_path)
+
         if not self.task:
-            self.task = self.infer_task_from_model_architecture(model_id_or_path)
+            self.task = self.infer_task_from_model_architecture(model_config)
+
+        with init_empty_weights():
```
Comment (johnugeorge) [inline python/huggingfaceserver/huggingfaceserver/model.py:120]: Can you add a comment here?

Comment (gavrissh) [reply python/huggingfaceserver/huggingfaceserver/model.py:120]: included

File: python/huggingfaceserver/huggingfaceserver/model.py [hunk 4]
```diff
@@ -108,31 +109,46 @@ def load(self) -> bool:
...
+            self.task = self.infer_task_from_model_architecture(model_config)
+
+        # device_map = "auto" enables model parallelism but all model architcture dont support it.
+        # For pre-check we initialize the model class without weights to check the `_no_split_modules`
+        # device_map = "auto" for models that support this else set to either cuda/cpu
+        with init_empty_weights():
+            self.model = AutoModel.from_config(model_config)
+
+        if self.model._no_split_modules:
+            self.device_map = "auto"
```
Comment (yuzisun) [inline python/huggingfaceserver/huggingfaceserver/model.py:124]: should we call `infer_auto_device_map` here?

Comment (yuzisun) [reply python/huggingfaceserver/huggingfaceserver/model.py:124]: https://huggingface.co/docs/accelerate/en/concept_guides/big_model_inference

Comment (gavrissh) [reply python/huggingfaceserver/huggingfaceserver/model.py:124]: I think the functionality wise it's the same

**Pull request merged**